File: Iodine/Toffee/macOS/Documents/FooDocument.java
package Documents;

import AppKit.*;

@objc public class FooDocument extends NSDocument {

	//
	// NSDocument implementation
	//

	@Override this withType(NSString type) error(NSError *error) { // workaround for invalid NSError pointer being passed in, sometimes
		data = new FooDocumentData();
	}

	@Override public boolean readFromURL(NSURL! url) ofType(NSString! typeName) error(__ref NSError error) {
		return loadDocumentFromURL(url);
	}

	@Override public boolean writeToURL(NSURL! url) ofType(NSString! typeName) forSaveOperation(NSSaveOperationType aveOperation) originalContentsURL(NSURL? absoluteOriginalContentsURL) error(__ref NSError error) {
		return saveDocumentToURL(url);
	}

	@Override public void makeWindowControllers() {
		addWindowController(new FooDocumentWindowController withDocument(self));
	}

	@Override public static boolean autosavesInPlace {
		__get {
			return true; // enables the title bar menu, among other things
		}
	}

	//
	// Custom members
	//

	boolean loadDocumentFromURL(NSURL url) {
		data = new FooDocumentData withURL(url);
		return data != null;
	}

	boolean saveDocumentToURL(NSURL url) {
		if (data == null)
			return false;
		data.save(url);
		return true;
	}

	public FooDocumentData data { __get; private __set; }
}

//
// data stored in the document:
//

public class FooDocumentData {

	public NSString text { __get; __set; }

	public this() {
		text = "New Foo!";
	}

	public this withURL(NSURL url) {
		if (File.Exists(url.path))
			text = File.ReadText(url.path);
		else
			return null;
	}

	public void save(NSURL url) {
		File.WriteText(url.path, text);
	}
}

File: Silver/Toffee/macOS/Documents/FooDocument.swift
import AppKit

@objc public class FooDocument : NSDocument {

	//
	// NSDocument implementation
	//

	override init(type: String?, error: UnsafePointer<NSError>) { // workaround for invalid NSError pointer being passed in, sometimes
		data = FooDocumentData()
	}

	override func read(from url: URL, ofType typeName: String) throws {
		try! loadDocument(from: url)
	}

	override func write(to url: URL, ofType typeName: String, `for` saveOperation: NSSaveOperationType, originalContentsURL absoluteOriginalContentsURL: URL?) throws {
		try! saveDocument(to: url)
	}

	override func makeWindowControllers() {
		addWindowController(FooDocumentWindowController(document: self))
	}

	override class var autosavesInPlace: Bool {
		return true // enables the title bar menu, among other things
	}

	//
	// Custom members
	//

	func loadDocument(from url: URL) throws {
		data = FooDocumentData(from: url)
		if data == nil {
			throw Exception("Count not load document.")
		}
	}

	func saveDocument(to url: URL) throws {
		if let data = data {
			data.save(to: url)
		}
		throw Exception("Count not save document.")
	}

	public var data: FooDocumentData?

}

//
// data stored in the document:
//

public class FooDocumentData {

	var text: String;

	init() {
		text = "New Foo!"
	}

	init?(from url: URL) {
		if File.Exists(url.path) {
			text = File.ReadText(url.path)
		} else {
			return nil
		}
	}

	func save(to url: URL) {
		File.WriteText(url.path, text)
	}

}